src/package-hoister.js

```diff
@@ -142,15 +146,15 @@ export default class PackageHoister {
     //
     let parentParts: Parts = [];
     const isIncompatible = ref.incompatible;
-    let isRequired = isDirectRequire && !ref.ignore && !isIncompatible;
+    let isRequired = isDirectRequire && !ref.ignore && !isIncompatible && (!ref.optional || !this.ignoreOptional);
```

Comment:
Not really stylistic. One reads "if `ref` is not optional or we should not ignore optionals" and the other one reads "if NOT (ref is optional and we should ignore optionals)". Human brains are bad at double negations ;)